### PR TITLE
Add scanAsync to Akka Streams, similar to scan but taking a Future

### PR DESCRIPTION
**Pull request by mateusduboli**
Refers to #21541

### CODE REVIEW

**Comment by typesafehub-validator** [comment]
Hi @mateusduboli,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by ktoso** [comment]
OK TO TEST

**Comment by mateusduboli** [comment]
I'm confused about the error message from CI, it states:
```
[error] /localhome/jenkinsakka/workspace/pr-validator-per-commit-jenkins/akka-docs/rst_preprocessed/scala/testing.rst:582: ERROR: Snippet (code/docs/testkit/ParentChildSpec.scala##test-TestProbe-parent) not found!
```
Not sure how to fix this :(

**Comment by mateusduboli** [comment]
@ktoso Can someone review this?

**Comment by ktoso** [comment]
Sure, sorry for the delay - was traveling and at customers for the last week :)

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: I find this name peculiar, how is the aggregated value an aggregator? What about `current`?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Not sure it makes sense to apply supervision to breaking streams complience. Should probably just throw/fail.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Will this actually ever be true, we just pushed to out it won't be available?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Doesn't this mean we push a value that may have already been pushed if resume is used?

**Review by johanandren — COMMENTED** [review]
There is a `stages-overview.rst` in the docs that needs to be updated for this as well.

Looks promising, great work!

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala
Comment (mateusduboli) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: There's the same behaviour in [`foldAsync`](https://github.com/mateusduboli/akka/blob/wip-scan-async-akka-stream/akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala#L569-L571).

I agree with you but this should be consistent with other operations.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala
Comment (mateusduboli) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Code speak better than words :)
Here are the [tests](https://github.com/mateusduboli/akka/blob/wip-scan-async-akka-stream/akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowScanAsyncSpec.scala#L102-L116)

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala
Comment (mateusduboli) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: This is actually wrong, I only need to check the `in` after the callback completes

**Comment by mateusduboli** [comment]
On the documentation I did for `foldAsync` too since it was missing.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala
Comment (mateusduboli) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: I tried to remove this, but the test for Large streams fails with this exception:
```
[info]   The future returned an exception of type: java.lang.IllegalArgumentException, with message: requirement failed: Cannot push port (ScanAsync.out) twice. (FlowScanAsyncSpec.scala:51)
```

And also the concrete value starts to become random

Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Ah, so if out has been closed we want to stop pulling, got it. The updated method looks better.

**Review by johanandren — APPROVED** [review]
LGTM

**Comment by mateusduboli** [comment]
@johanandren According to [CONTRIBUTING.md](https://github.com/akka/akka/blob/master/CONTRIBUTING.md) I need 2 LGTMs to get this merged.
I'm waiting this before I squash my commits onto one with a nice message, should I wait or just get this done?

**Comment by johanandren** [comment]
No need for you to squash it as github has support for that now, it needs another review before we can merge it though.

**Comment by johanandren** [comment]
There seems to be a compilation failure now that needs to be resolved as well.

**Comment by mateusduboli** [comment]
@ktoso @johanandren Can anyone else review this PR? It's been almost a week with no activity :(

**Comment by johanandren** [comment]
Sorry @mateusduboli more things to do than CPU cycles I'm afraid. We'll look into it as soon as we can!

**Review by patriknw — APPROVED** [review]
LGTM, thanks @mateusduboli

**Comment by ktoso** [comment]
Hi @mateusduboli, for future reference - operations also need docs in javadsl, we missed this in the review process (sorry, our fault).
I added the missing docs here: https://github.com/akka/akka/pull/21701

**Comment by mateusduboli** [comment]
@ktoso I took the freedom to make a PR adding this step in `CONTRIBUTING.md`, I think that would have helped me reviewing the docs first :)
https://github.com/akka/akka/pull/21703

**Pull request merged**
